api/base/parsers.py

```diff
@@ -15,6 +17,29 @@ class JSONAPIParser(JSONParser):
     media_type = 'application/vnd.api+json'
     renderer_class = JSONAPIRenderer
 
+    def flatten_relationships(self, relationships):
+        """
+        Flattens relationships dictionary which has information needed to create related resource objects.
+
+        Validates that formatting of relationships dictionary is correct.
+        """
+        if not isinstance(relationships, dict):
+                raise ParseError()
+        related_resource = relationships.keys()[0]
+        if not isinstance(relationships[related_resource], dict) or related_resource == 'data':
+            raise ParseError()
+        data = relationships[related_resource].get('data')
+
+        if not data:
+            raise JSONAPIException(source={'pointer': '/relationships/<related_resource_name>/data'}, detail=NO_DATA_ERROR)
```

Comment:
Not saying this is definitely the right solution, but I have some pending changes that might streamline this code:
https://github.com/CenterForOpenScience/osf.io/pull/4246/files#diff-2d13bee79720b482d2d8219a37b07c9aR83